src/dvc_data/diff.py

```diff
@@ -14,43 +15,25 @@
 UNCHANGED = "unchanged"
 
 
+@define
 class TreeEntry:
-    __slots__ = ("in_cache", "key", "meta", "oid")
-
-    def __init__(
-        self,
-        in_cache: bool,
-        key: Tuple[str],
-        meta: Optional["Meta"],
-        oid: Optional["HashInfo"],
-    ):
-        self.in_cache = in_cache
-        self.key = key
-        self.meta = meta
-        self.oid = oid
+    in_cache: bool = field(default=False, eq=False)
+    key: Tuple[str, ...] = ()
+    meta: Optional["Meta"] = field(default=None, eq=False)
+    oid: Optional["HashInfo"] = None
 
     def __bool__(self):
         return bool(self.oid)
 
-    def __eq__(self, other):
-        if not isinstance(other, TreeEntry):
-            return False
-
-        if self.key != other.key:
-            return False
-
-        return self.oid == other.oid
-
 
+@define
 class Change:
-    __slots__ = ("old", "new")
-
-    def __init__(self, old: TreeEntry, new: TreeEntry):
-        self.old = old
-        self.new = new
+    old: TreeEntry = field(factory=TreeEntry)
+    new: TreeEntry = field(factory=TreeEntry)
+    typ: str = field(init=False)
 
-    @property
-    def typ(self):
+    @typ.default
+    def _(self):
```

Comment:
With this, `typ` is generated when the instance is created rather than lazily evaluating.
We call `typ` at the end anyway, so I think this is better than evaluating it multiple times during `diff`. Also I think `typ` is a property of `Change`, so I think it makes sense to tie this to object instantiation itself. It's a derived property.